src/System.Management.Automation/engine/CoreAdapter.cs

```diff
@@ -1693,13 +1696,19 @@ internal static Type EffectiveArgumentType(object arg)
                         return firstType.MakeArrayType();
                     }
                 }
-
-                return arg.GetType();
             }
-            else
+
+            return GetObjectType(arg, skipBasing: true);
+        }
+
+        internal static Type GetObjectType(object obj, bool skipBasing)
```

Comment:
1. Is `basing` a word? `skipBaseResolution` sounds better to me.
2. Please add a comment why you are resolving the base exactly once, instead of e.g. all the way to root.
3. Finally, is this base resolution even reasonable inside this method? `GetObjectType` does not sound like anything that should deal with hierarchy, only the current object type. If caller wants to use the base type, they should pass `PSObject.Base(obj)` themselves explicitly.

Reply:
> * Please add a comment why you are resolving the base exactly once, instead of e.g. all the way to root.

`PSObject.Base` gets the first non-`PSObject` base object. In other words, it does base all the way to the root.

> * Finally, is this base resolution even reasonable inside this method? `GetObjectType` does not sound like anything that should deal with hierarchy, only the current object type. If caller wants to use the base type, they should pass `PSObject.Base(obj)` themselves explicitly.

`PSObject` is a wrapper that helps with engine implementation details (and it's a property bag). It's not related to hierarchy.

Reply:
It turns out we already use the word [`debase`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/LanguagePrimitives.cs#L4761) to indicate if `PSObject.Base` should be called. Will update to follow the existing naming and pattern.